File: src/MCCore/MCCore.cpp
#include <filesystem>
#include <set>

#include "file/file.hpp"
#include "MCCore.hpp"
#include "network/network.hpp"
#include "strTools/strTools.hpp"

using sonic_json::Document;
using std::cout;
using std::endl;
using std::filesystem::exists;
using std::set;
using std::to_string;

// 已有变量：gameDir - .minecraft位置    javaDir - java位置(可执行文件)

#ifdef _WIN32
    string os = "windows";
#elif __APPLE__
    string os = "osx";
#elif __linux__
    string os = "linux";
#else
    string os = ""
    #Warning mc_core: Your OS or compiler may not be supported.
#endif

#ifdef __aarch64__
    string arch = "aarch64";
#elif __arm__
    string arch = "arm";
#elif __x86_64__
    string arch = "x86_64";
#elif __i386__
    string arch = "x86";
#else
    string arch = "";
    #Warning mc_core: Your OS or compiler may not be supported.
#endif

string addArgs(Node &n) {
    string result = "";
    for (int i = 0; n.AtPointer(i) != nullptr; i++) {
        if (n.AtPointer(i)->IsString()) {
            // 无限制，直接添加
            result.append(n.AtPointer(i)->GetString()).append(" ");
            continue;
        }

        // 不是String，判断rules
        if (!n.AtPointer(i)->HasMember("value")) continue; // 格式错误
        if (!n.AtPointer(i)->HasMember("rules")) continue; // 格式错误
        Node * r = n.AtPointer(i, "rules");
        bool allow = true;
        for (int j = 0; r->AtPointer(j) != nullptr; j++) {
            if (!r->AtPointer(j)->HasMember("action")) {
                allow = false;
                break;
            }
            if (r->AtPointer(j)->HasMember("features")) {
                allow = false;
                break; // not support yet
            }
            if (r->AtPointer(j, "action")->GetString() == "allow") {
                if (r->AtPointer(j, "os")->HasMember("name") &&
                    r->AtPointer(j, "os", "name")->GetString() != os) {
                    allow = false;
                    break;
                }
                if (r->AtPointer(j, "os")->HasMember("arch") &&
                    r->AtPointer(j, "os", "arch")->GetString() != arch) {
                    allow = false;
                    break;
                }
            } else if (r->AtPointer(j, "action")->GetString() == "disallow") {
                if (r->AtPointer(j, "os")->HasMember("name") &&
                    r->AtPointer(j, "os", "name")->GetString() == os) {
                    allow = false;
                    break;
                }
                if (r->AtPointer(j, "os")->HasMember("arch") &&
                    r->AtPointer(j, "os", "arch")->GetString() == arch) {
                    allow = false;
                    break;
                }
            }
        }
        
        if (!allow) continue;
        if (n.AtPointer(i, "value")->IsString()) {
            // 是一条参数
            result.append(n.AtPointer(i, "value")->GetString()).append(" ");
        } else {
            // 是数组
            for (int k = 0; n.AtPointer(i, "value", k) != nullptr; k++) {
                if (!n.AtPointer(i, "value", k)->IsString()) continue; // 格式错误
                result.append(n.AtPointer(i, "value", k)->GetString()).append(" ");
            }
        }
    }
    return result;
}

bool addGame() {
    return true;
}

// TODO 下载相关
vector<string> addLibs(Node &n, string gameDir) {
    vector<string> result;
    for (int i = 0; n.AtPointer(i) != nullptr; i++) {
        if (!n.AtPointer(i)->HasMember("name")) continue; // 格式错误
        string temp = gameDir + "/libraries/";
        bool allow = true;
        if (n.AtPointer(i)->HasMember("rules")) {
            // 有rules，进行判断
            Node * r = n.AtPointer(i, "rules");
            for (int j = 0; r->AtPointer(j) != nullptr; j++) {
                if (!r->AtPointer(j)->HasMember("action")) {
                    allow = false;
                    break;
                }
                // ver < 1.13 存在{"action":"allow"}, etc.
                if (!r->AtPointer(j)->HasMember("os")) continue;
                if (r->AtPointer(j, "action")->GetString() == "allow") {
                    if (r->AtPointer(j, "os")->HasMember("name") &&
                        r->AtPointer(j, "os", "name")->GetString() != os) {
                        allow = false;
                        break;
                    }
                    if (r->AtPointer(j, "os")->HasMember("arch") &&
                        r->AtPointer(j, "os", "arch")->GetString() != arch) {
                        allow = false;
                        break;
                    }
                } else if (r->AtPointer(j, "action")->GetString() == "disallow") {
                    if (r->AtPointer(j, "os")->HasMember("name") &&
                        r->AtPointer(j, "os", "name")->GetString() == os) {
                        allow = false;
                        break;
                    }
                    if (r->AtPointer(j, "os")->HasMember("arch") &&
                        r->AtPointer(j, "os", "arch")->GetString() == arch) {
                        allow = false;
                        break;
                    }
                }
            }
        }
        if (!allow) continue;
        // name: 包名:jar名:版本
        vector<string> nameS = splitStr(n.AtPointer(i, "name")->GetString(), ':');
        if (nameS.size() != 3) continue; // 格式错误
        strReplace(&nameS[0], ".", "/"); // 包名转换路径
        temp.append(nameS[0])
            .append("/")
            .append(nameS[1])
            .append("/")
            .append(nameS[2])
            .append("/")
            .append(nameS[1])
            .append("-")
            .append(nameS[2])
            .append(".jar");
        result.push_back(temp);
    }
    return result;
}

Account addOfflineAccount() {
    Account account;
    return account;
}

Account addOnlineAccount() {
    Account account;
    return account;
}

bool delGame() {
    return true;
}

string getCMD(Account account, Game game, string javaDir, string gameDir) {
    #ifdef DEBUG
        cout << "[Info] mc_core::getCMD : Start." << endl;
    #endif
    // 使用自定义参数：
    if (!game.args.empty()) return javaDir + " " + game.args;    

    #ifdef DEBUG
        cout << "[Info] mc_core::getCMD : OS Type:" << os << ", " << arch << endl;
    #endif

    // 生成参数：

    string result = javaDir + " "; // 结果
    string jvmArg = string("-XX:+UseG1GC ") +
                    "-XX:-UseAdaptiveSizePolicy " +
                    "-XX:-OmitStackTraceInFastThrow " +
                    "-Dfml.ignoreInvalidMinecraftCertificates=True " +
                    "-Dfml.ignorePatchDiscrepancies=True " + 
                    "-Dlog4j2.formatMsgNoLookups=true "; // JVM参数
    string gameArg = ""; // 游戏参数
    vector<string> cp; // -cp内容
    string mainClass = ""; // mainClass值
    string assetIndex = ""; // assetIndex值

    // TODO 添加操作系统信息

    Document doc;
    doc.Parse(openFile(gameDir + "/versions/" + game.version + "/" + game.version + ".json"));
    if (doc.HasParseError()) {
        cout << "[Error] mc_core::getCMD : Failed to get launch command: "
             << gameDir + "/versions/" + game.version + "/" + game.version + ".json"
             << ": in line" << doc.GetErrorOffset() << ": Parse Error." << endl;
        return "";
    }

    // 判断inheritsFrom
    if (doc.HasMember("inheritsFrom")) {
        string parentVer = doc.AtPointer("inheritsFrom")->GetString();
        if (exists(gameDir + "/versions/" + parentVer)) {
            Document par;
            par.Parse(openFile(gameDir + "/versions/" + parentVer + "/" + parentVer + ".json"));
            if (par.HasParseError()) {
                cout << "[Error] mc_core::getCMD : Failed to get launch command: "
                     << gameDir + "/versions/" + parentVer + "/" + parentVer + ".json"
                     << ": in line" << par.GetErrorOffset() << ": Parse Error." << endl;
                return "";
            }
            mainClass = doc.AtPointer("mainClass")->GetString() + " ";
            assetIndex = par.AtPointer("assetIndex", "id")->GetString() + " ";
            cp = addLibs(*par.AtPointer("libraries"), gameDir);
            cp.push_back(gameDir + "/versions/" + parentVer + "/" + parentVer + ".jar:");
            if (!(setArgs(par, &jvmArg, &gameArg) && setArgs(doc, &jvmArg, &gameArg))) return "";
        } else {
            // TODO 下载原版
            return "";
        }
    } else {
        if (!setArgs(doc, &jvmArg, &gameArg)) return "";
        assetIndex = doc.AtPointer("assetIndex", "id")->GetString();
        mainClass = doc.AtPointer("mainClass")->GetString() + " ";
    }

    result.append(jvmArg) // JVM
          .append("${authlib_injector_param} -Xms" + game.xms + " -Xmx" + game.xmx + " ") // 额外JVM
          .append(mainClass) // 主类
          .append(gameArg) // 游戏参数
          .append("--width " + to_string(game.width) + " --height " + to_string(game.height) + " "); // 额外游戏参数

    // Optifine : --tweakClass调整至末尾
    if (result.find("--tweakClass optifine.OptiFineTweaker") != result.npos) {
        strReplace(&result, "--tweakClass optifine.OptiFineTweaker", "");
    }
    if (result.find("--tweakClass optifine.OptiFineForgeTweaker") != result.npos) {
        strReplace(&result, "--tweakClass optifine.OptiFineForgeTweaker", "");
    }

    // 替换result中的模板

    // ${classpath}
    string cps = "";
    vector<string> cp2 = addLibs(*doc.AtPointer("libraries"), gameDir);
    cp.insert(cp.end(), cp2.begin(), cp2.end());
    cp.push_back(gameDir + "/versions/" + game.version + "/" + game.version + ".jar");
    // 去重
    set<string> s(cp.begin(), cp.end());
    cp.assign(s.begin(), s.end());
    for (string s : cp) {
        cps.append(s).append(":");
    }

    strReplace(&result, "${auth_player_name}", account.userName);
    strReplace(&result, "${version_name}", game.version);
    strReplace(&result, "${game_directory}", gameDir);
    strReplace(&result, "${assets_root}", gameDir + "/assets");
    strReplace(&result, "${assets_index_name}", assetIndex);
    strReplace(&result, "${auth_uuid}", account.uuid);
    strReplace(&result, "${auth_access_token}", account.token);
    strReplace(&result, "${user_type}", account.type);
    strReplace(&result, "${version_type}", game.type);
    strReplace(&result, "${natives_directory}", gameDir + "/versions/" + game.version + "/natives");
    strReplace(&result, "${launcher_name}", "\"CE Minecraft Launcher\"");
    strReplace(&result, "${launcher_version}", "1.0.0");
    strReplace(&result, "${classpath}", cps);
    strReplace(&result, "${library_directory}", gameDir + "/libraries");
    strReplace(&result, "${classpath_separator}", ":");
    strReplace(&result, "${authlib_injector_param}", ""); // not support yet

    // TODO natives

    cout << result << endl;

    #ifdef DEBUG
        cout << "[Info] mc_core::getCMD : Finished. Command: " << result << endl;
    #endif
    return result;
}

vector<Game> loadGameList(
    bool reload,
    string gameDir,
    int defHeight,
    int defWidth,
    string defXms,
    string defXmx) {
    vector<Game> gameList;
    if (reload) {
        if (!exists(gameDir + "/versions")) return {};
        vector<string> versionList = getDirs(gameDir + "/versions");
        int count = versionList.size();
        gameList.resize(count);
        for (int i = 0; i < count; i++) {
            // Get version type
            Document doc;
            doc.Parse(openFile(gameDir + "/versions/" + versionList[i] + "/" + versionList[i] + ".json"));
            if (doc.HasParseError()) {
                cout << "[Error] mc_core::loadGameList : Failed to load " << gameDir + "/versions/" + versionList[i] + "/" + versionList[i] + ".json"
                     << ": in line" << doc.GetErrorOffset() << ": has parse error." << endl;
                continue;
            }
            if (!doc.HasMember("type")) {
                cout << "[Error] mc_core::loadGameList : Failed to load " << gameDir + "/versions/" + versionList[i] + "/" + versionList[i] + ".json"
                     << ": doesn't have member \"type\"" << endl;
                continue;
            }

            gameList[i].args = "";
            gameList[i].height = defHeight;
            gameList[i].type = doc.AtPointer("type")->GetString();
            gameList[i].version = versionList[i];
            gameList[i].width = defWidth;
            gameList[i].xms = defXms;
            gameList[i].xmx = defXmx;
        }
        // TODO SHA-1 sum
    } else {
        string cfgText = openFile("index.json");
        if (cfgText.empty()) {
            return loadGameList(true, gameDir, defHeight, defWidth, defXms, defXmx);
        }
        Document doc;
        doc.Parse(cfgText);
        // TODO: if SHA-1 different, reload
        if (doc.HasParseError()) {
            return loadGameList(true, gameDir, defHeight, defWidth, defXms, defXmx);
        }

        Node * list = doc.AtPointer("gameList");
        for (int i = 0; list->AtPointer(i) != nullptr; i++) {
            Game g;
            g.args = list->AtPointer(i, "args")->GetString();
            g.height = list->AtPointer(i, "height")->GetInt64();
            g.type = list->AtPointer(i, "type")->GetString();
            g.version = list->AtPointer(i, "version")->GetString();
            g.width = list->AtPointer(i, "width")->GetInt64();
            g.xms = list->AtPointer(i, "xms")->GetString();
            g.xmx = list->AtPointer(i, "xmx")->GetString();
            gameList.push_back(g);
        }
    }
    return gameList;
}

bool setArgs(Node &n, string *jvmArg, string *gameArg) {
    if (!n.HasMember("arguments") && n.HasMember("minecraftArguments")) {
        // ver < 1.13
        jvmArg->append("-Djava.library.path=${natives_directory} -cp ${classpath} ");
        gameArg->append(n.AtPointer("minecraftArguments")->GetString());
    } else if (n.HasMember("arguments") && n.AtPointer("arguments")->HasMember("jvm") && n.AtPointer("arguments")->HasMember("game")) {
        // ver >= 1.13
        jvmArg->append(addArgs(*n.AtPointer("arguments", "jvm")));
        gameArg->append(addArgs(*n.AtPointer("arguments", "game")));
    } else {
        // 错误
        cout << "[Error] mc_core::setArgs : Failed to get jvm and game arguments: Format Error." << endl;
        return false;
    }
    return true;
}

File: src/mc_core.rs
use log::{info, warn, error};
use serde_json::Value;
use std::{env::consts as env, fs::read_to_string};

use crate::{file_tools::{exists, list_dir}, Config};

// 游戏账号
pub struct Account {
    // 登录类型，直接填入
    pub account_type: String,
    // access_token,直接填入
    pub token: String,
    // uuid，直接填入
    pub uuid: String,
    // user_name，直接填入
    pub user_name: String,
}

// 游戏配置
pub struct Game {
    // 自定义参数，留空则使用默认
    pub args: String,
    // 描述，由用户输入
    pub description: String,
    // 窗口高度
    pub height: isize,
    // java可执行文件路径
    pub java_path: String,
    // 启用版本隔离
    pub seperated: bool,
    // 游戏类型，直接填入
    pub game_type: String,
    // 游戏版本，直接填入
    pub version: String,
    // 窗口宽度
    pub width: isize,
    // xms参数
    pub xms: String,
    // xmx参数
    pub xmx: String,
}

// 获取单条参数
fn add_arg(n: &Value) -> Option<Vec<String>> {
    let mut result: Vec<String> = Vec::new();

    for item in n.as_array()? {
        if item.is_string() {
            // 无限制，可直接添加
            result.push(item.as_str().unwrap().into());
            continue;
        }
        
        // 判断是否满足限制条件
        if !check_rules(&item["rules"]) {
            continue;
        }

        if item["value"].is_string() {
            // 单条参数
            result.push(item["value"].as_str().unwrap().into());
        } else {
            // 数组
            for arg in item["value"].as_array()? {
                result.push(arg.as_str()?.into());
            }
        }
    }

    Some(result)
}

// 检查参数是否可以添加
fn check_rules(n: &Value) -> bool {
    // 获取操作系统名称
    let os: &str;
    if env::OS == "macOS" {
        os = "osx";
    } else {
        os = env::OS;
    }

    if let Some(array) = n.as_array() {
        for r in array {
            if !r["features"].is_null() {
                // 暂时不支持
                return false;
            }
            if r["action"] == "allow" {
                if r["os"]["arch"] != env::ARCH {
                    return false;
                }
                if r["os"]["name"] != os {
                    return false;
                }
            } else if r["action"] == "disallow" {
                if r["os"]["arch"] == env::ARCH {
                    return false;
                }
                if r["os"]["name"] == os {
                    return false;
                }
            }
        }
    } else {
        warn!("Failed to get rules");
    }
    true
}

// 获取MC和JVM参数
fn get_args(n: &Value) -> Option<(Vec<String>, Vec<String>)> {
    let mut game_args:Vec<String> = Vec::new();
    let mut jvm_args:Vec<String> = vec![
        "-XX:+UseG1GC".to_string(),
        "-XX:-UseAdaptiveSizePolicy".to_string(),
        "-XX:-OmitStackTraceInFastThrow".to_string(),
        "-Dfml.ignoreInvalidMinecraftCertificates=True".to_string(),
        "-Dfml.ignorePatchDiscrepancies=True".to_string(),
        "-Dlog4j2.formatMsgNoLookups=true".to_string()
    ];

    if !n["arguments"].is_null() {
        // MC版本 >= 1.13
        game_args.append(&mut add_arg(&n["arguments"]["game"])?);
        jvm_args.append(&mut add_arg(&n["arguments"]["jvm"])?);
    } else {
        // MC版本 < 1.13
        let args: Vec<&str> = n["minecraftArguments"].as_str()?.split(" ").collect();
        for arg in args {
            game_args.push(arg.into());
        }
        jvm_args.append(&mut vec![
            "-Djava.library.path=${natives_directory}".into(), 
            "-cp".into(),
            "${classpath}".into()
        ]);
    }

    Some((game_args, jvm_args))
}

// 获取-cp参数 
fn get_classpaths(n: &Value, game_path: &String) -> Option<Vec<String>> {
    let mut result: Vec<String> = Vec::new();
    for item in n.as_array()? {
        let mut temp: String = game_path.clone() + "/libraries/";

        if !item["rules"].is_null() &&
            check_rules(&item["rules"]) {
            continue;
        }

        let name = String::from(item["name"].as_str()?);
        let name_split: Vec<&str> = name.split(":").collect();
        temp.push_str((
            name_split[0].replace(".", "/") + "/" +
            name_split[1] + "/" +
            name_split[2] + "/" +
            name_split[1] + "-" + name_split[2]
        ).as_str());
        if name_split.len() == 4 {
            // 添加后缀
            temp.push_str(("-".to_string() + name_split[3]).as_str());
        }
        temp.push_str(".jar");

        result.push(temp);
    }

    Some(result)
}

// 获取启动总命令
pub fn get_launch_command(account: &Account, game: &Game, game_path: &String) -> Option<Vec<String>> {
    info!("Start.");
    // 使用自定义参数
    // if !game.args.is_empty() {
    //     return game.args.clone();
    // }
    let mut result: Vec<String> = Vec::new();
    let dir = game_path.clone() + "/versions/" + game.version.as_str();  // 游戏目录
    
    // TODO: 或可与启动时的load_game_list合并
    // 读取json
    let cfg_path = dir.clone() + "/" + game.version.as_str() + ".json";
    if let Ok(config) = serde_json::from_str::<Value>(read_to_string(&cfg_path).ok()?.as_str()) {
        // assetIndex
        let asset_index: String;
        // forge需要提前写入的参数
        let mut classpaths: Vec<String> = Vec::new();
        let mut game_args:Vec<String> = Vec::new();
        let mut jvm_args:Vec<String> = Vec::new();
        // 判断inheritsFrom（forge需要）
        if config["inheritsFrom"].is_null() {
            // 无forge
            if let Some((temp_game_args, temp_jvm_args)) = get_args(&config) {
                game_args = temp_game_args;
                jvm_args = temp_jvm_args;
                if let Some(index) = config["assetIndex"]["id"].as_str() {
                    asset_index = index.into();
                } else {
                    error!("Failed to get assetIndex.");
                    return None;
                }
            } else {
                error!("Failed to get game arguments and jvm arguments.");
                return None;
            }
        } else {
            // 有forge
            if let Some(parent_version) = config["inheritsFrom"].as_str() {
                let parent_path = game_path.clone() + "/versions/" + &parent_version;
                if exists(&parent_path) {
                    if let Ok(parent) = serde_json::from_str::<Value>(&read_to_string(&parent_path).ok()?.as_str()) {
                        if let Some(index) = parent["assetIndex"]["id"].as_str() {
                            asset_index = index.into();
                        } else {
                            error!("Failed to get assetIndex.");
                            return None;
                        }
                        // MC和JVM的参数
                        if let (
                            Some((mut parent_game_args, mut parent_jvm_args)),
                            Some((mut self_game_args, mut self_jvm_args))
                        ) = (get_args(&parent), get_args(&config)) {
                            game_args.append(&mut parent_game_args);
                            game_args.append(&mut self_game_args);
                            jvm_args.append(&mut parent_jvm_args);
                            jvm_args.append(&mut self_jvm_args);
                        } else {
                            error!("Failed to get arguments from {cfg_path}.");
                            return None;
                        }
                        // classpaths列表
                        if let Some(vector) = get_classpaths(&parent["libraries"], game_path) {
                            classpaths = vector;
                        } else {
                            error!("Failed to load classpaths.");
                            return None;
                        }
                    } else {
                        error!("Failed to load {parent_path}.");
                        return None;
                    }
                } else {
                    // TODO: 下载原版
                    return None;
                }
            } else {
                error!("Failed to get inheritsFrom.");
                return None;
            }
        }

        // classpaths列表
        if let Some(mut vector) = get_classpaths(&config["libraries"], game_path) {
            classpaths.append(&mut vector);
        } else {
            error!("Failed to load classpaths.");
            return None;
        }
        classpaths.push(dir.clone() + "/" + game.version.as_str() + ".jar");  // 游戏本身
        
        // classpaths列表去重，获得最终字符串
        let mut i = 0;
        let mut cp = String::new();
        let l = classpaths.len();
        while i < l {
            if !classpaths[i+1..l].contains(&classpaths[i]) {
                cp.push_str((classpaths[i].clone() + ":").as_str());
            }
            i += 1;
        }

        // 设置额外参数 TODO: 更多自定义
        jvm_args.append(&mut vec![
            /*"${authlib_injector_param}".into(), */
            "-Xms".to_string() + &game.xms,
            "-Xmx".to_string() + &game.xmx
        ]);
        game_args.append(&mut vec![
            "--height".into(),
            game.height.to_string(),
            "--width".into(),
            game.width.to_string()
        ]);

        // 参数添加至result
        result.append(&mut jvm_args);
        // 主类
        if let Some(main_class) = config["mainClass"].as_str() {
            result.push(main_class.to_string());
        } else {
            error!("Failed to get mainClass.");
            return None;
        }
        result.append(&mut game_args);

        // 替换模板
        for item in result.iter_mut() {
            // TODO: 优化替换
            *item = item
                .replace("${assets_index_name}", &asset_index)
                .replace("${assets_root}", &(game_path.clone() + "/assets"))
                .replace("${auth_access_token}", &account.token)
                .replace("${auth_player_name}", &account.user_name)
                .replace("${auth_uuid}", &account.uuid)
                // .replace("${authlib_injector_param}", "") // 暂不支持
                .replace("${classpath}", &cp)
                .replace("${classpath_separator}", ":")
                .replace("${game_directory}", &game_path)
                .replace("${launcher_name}", "\"CE Minecraft Launcher\"")
                .replace("${launcher_version}", env!("CARGO_PKG_VERSION"))
                .replace("${library_directory}", &(game_path.clone() + "/libraries"))
                .replace("${natives_directory}", &(dir.clone() + "/natives"))
                .replace("${user_type}", &account.account_type)
                .replace("${version_name}", &game.version)
                .replace("${version_type}", &game.game_type);
        }
        return Some(result)
    } else {
        error!("Failed to load {cfg_path}.");
        return None;
    }
}

pub fn load_game_list(config: &Config) -> Option<Vec<Game>> {
    let mut game_list: Vec<Game> = Vec::new();
    let dir = config.game_path.borrow().clone() + "/versions";

    if !exists(&dir) {
        // 空目录
        warn!("{dir} is empty.");
        return Some(game_list);
    }

    for version in list_dir(&dir)? {
        let mut game: Game;
        let path = dir.clone() + "/" + version.as_str();
        
        // 先加载原版json
        if let Ok(json) = serde_json::from_str::<Value>(&read_to_string(&(path.clone() + "/" + &version.as_str() + ".json")).ok()?.as_str()) {
            game = Game {
                args: "".into(),
                description: "".into(),
                height: config.height.borrow().clone(),
                java_path: config.java_path.borrow().clone(),
                seperated: false,
                game_type: json["type"].as_str()?.into(),
                version: version.into(),
                width: config.width.borrow().clone(),
                xms: config.xms.borrow().clone(),
                xmx: config.xmx.borrow().clone()
            };
        } else {
            // 异常，跳过此次加载
            warn!("Failed to load {version}.json.");
            continue;
        }
        
        // 若config.json存在，覆盖原版json
        let cfg_path = path.clone() + "/" + "config.json";
        if exists(&cfg_path) {
            if let Ok(json) = serde_json::from_str::<Value>(&read_to_string(&cfg_path).ok()?.as_str()) {
                for item in json.as_array()? {
                    game.args = item["args"].as_str()?.into();
                    game.description = item["description"].as_str()?.into();
                    game.height = item["height"].as_i64()? as isize;
                    game.java_path = item["java_path"].as_str()?.into();
                    game.seperated = item["seperated"].as_bool()?;
                    game.width = item["width"].as_i64()? as isize;
                    game.xms = item["xms"].as_str()?.into();
                    game.xmx = item["xmx"].as_str()?.into();
                }
            } else {
                warn!("Failed to load {cfg_path}.");
                continue;
            }
        }
        game_list.push(game);
    };
    Some(game_list)
}
